File: backend/src/main/java/com/bloodbridge/service/DeviceTokenService.java
package com.bloodbridge.service;

import com.bloodbridge.dto.request.RegisterDeviceTokenRequest;
import com.bloodbridge.dto.response.ApiResponse;
import com.bloodbridge.dto.response.DeviceTokenResponse;

import java.util.List;

/**
 * Service interface for managing Firebase Cloud Messaging device tokens.
 * Phase 3B.1 — Device Registration module.
 */
public interface DeviceTokenService {

    /**
     * Register or update an FCM device token for the authenticated user.
     * Upsert pattern: if the token already exists, update lastSeen and browser metadata without creating a duplicate.
     *
     * @param userEmail Email of the authenticated user
     * @param request   DTO containing token and device metadata
     * @return Standard ApiResponse with DeviceTokenResponse
     */
    ApiResponse<DeviceTokenResponse> registerToken(String userEmail, RegisterDeviceTokenRequest request);

    /**
     * Handle token rotation issued by Firebase SDK.
     * Replaces oldToken with newToken for the authenticated user.
     *
     * @param userEmail Email of the authenticated user
     * @param oldToken  Previous FCM token
     * @param newToken  Newly issued FCM token
     * @return Standard ApiResponse with DeviceTokenResponse
     */
    ApiResponse<DeviceTokenResponse> refreshToken(String userEmail, String oldToken, String newToken);

    /**
     * Deactivate or delete an FCM device token (e.g., on logout).
     *
     * @param userEmail Email of the authenticated user
     * @param fcmToken  FCM token string to remove
     * @return Standard ApiResponse
     */
    ApiResponse<Void> removeToken(String userEmail, String fcmToken);

    /**
     * Touch lastSeen timestamp for an FCM token.
     *
     * @param userEmail Email of the authenticated user
     * @param fcmToken  FCM token string
     * @return Standard ApiResponse
     */
    ApiResponse<Void> touchLastSeen(String userEmail, String fcmToken);

    /**
     * Retrieve all active device tokens for the authenticated user.
     *
     * @param userEmail Email of the authenticated user
     * @return Standard ApiResponse containing list of DeviceTokenResponse
     */
    ApiResponse<List<DeviceTokenResponse>> getUserActiveTokens(String userEmail);
}

File: frontend/src/services/deviceTokenService.js
import api from '../api/axios';

/**
 * Service for interacting with backend Device Token endpoints (/api/v1/device/*).
 * Phase 3B.1 — Device Registration module.
 */
export const deviceTokenService = {
  /**
   * Register or update an FCM device token with the backend.
   *
   * @param {Object} data - { token, platform, browser, deviceName, deviceId }
   * @returns {Promise<Object>} ApiResponse containing DeviceTokenResponse
   */
  registerDeviceToken: async (data) => {
    const response = await api.post('/device/register', data);
    return response.data?.data ?? response.data;
  },

  /**
   * Refresh an updated FCM token with the backend.
   *
   * @param {string} oldToken - Previous token (optional)
   * @param {string} newToken - New token
   * @returns {Promise<Object>} ApiResponse containing DeviceTokenResponse
   */
  refreshDeviceToken: async (oldToken, newToken) => {
    const params = new URLSearchParams();
    if (oldToken) params.append('oldToken', oldToken);
    params.append('newToken', newToken);

    const response = await api.post(`/device/refresh?${params.toString()}`);
    return response.data?.data ?? response.data;
  },

  /**
   * Remove/deactivate an FCM device token on the backend (e.g. on logout).
   *
   * @param {string} token - FCM token to remove
   * @returns {Promise<Object>} ApiResponse
   */
  removeDeviceToken: async (token) => {
    const response = await api.delete(`/device/remove?token=${encodeURIComponent(token)}`);
    return response.data;
  },

  /**
   * Fetch all active registered devices for the current user.
   *
   * @returns {Promise<Array>} List of registered device responses
   */
  getMyDevices: async () => {
    const response = await api.get('/device/my-devices');
    return response.data?.data ?? response.data;
  },
};

export default deviceTokenService;
